server/src/main/java/org/elasticsearch/env/NodeMetadata.java

```diff
@@ -138,36 +139,36 @@ public void verifyUpgradeToCurrentVersion() {
     public NodeMetadata upgradeToCurrentVersion() {
         verifyUpgradeToCurrentVersion();
 
-        return nodeVersion.equals(Version.CURRENT) ? this : new NodeMetadata(nodeId, Version.CURRENT, nodeVersion, oldestIndexVersion);
+        return nodeVersion.equals(BuildVersion.current())
+            ? this
+            : new NodeMetadata(nodeId, BuildVersion.current(), nodeVersion, oldestIndexVersion);
     }
 
     private static class Builder {
         String nodeId;
-        Version nodeVersion;
-        Version previousNodeVersion;
+        BuildVersion nodeVersion;
+        BuildVersion previousNodeVersion;
         IndexVersion oldestIndexVersion;
 
         public void setNodeId(String nodeId) {
             this.nodeId = nodeId;
         }
 
         public void setNodeVersionId(int nodeVersionId) {
-            this.nodeVersion = Version.fromId(nodeVersionId);
+            this.nodeVersion = BuildVersion.fromVersionId(nodeVersionId);
         }
 
         public void setOldestIndexVersion(int oldestIndexVersion) {
             this.oldestIndexVersion = IndexVersion.fromId(oldestIndexVersion);
         }
 
-        private Version getVersionOrFallbackToEmpty() {
-            return Objects.requireNonNullElse(this.nodeVersion, Version.V_EMPTY);
-        }
-
+        @UpdateForV9 // version is required in the node metadata from v9 onwards
         public NodeMetadata build() {
-            @UpdateForV9 // version is required in the node metadata from v9 onwards
-            final Version nodeVersion = getVersionOrFallbackToEmpty();
             final IndexVersion oldestIndexVersion;
 
+            if (this.nodeVersion == null) {
+                nodeVersion = BuildVersion.fromVersionId(0);
```

Comment:
This id would be newer than the fixed -1 serverless uses. Can we ensure these never conflict for serverless? For example, asserting that any version id passed to serverless is rejected if it isn't -1.

Reply:
We've got that assertion: https://github.com/elastic/elasticsearch-serverless/pull/1466/files#diff-a52b488e254cc7aece3b67f95436c3e733eb1794f997abf3f10b074c9815d5b6R114

I can add an assertion on this side that the DefaultBuildVersion is always created with a positive integer for its version ID.